### PR TITLE
Georgettica/bump external snapshotter version (fixes #2966)

### LINKED ISSUE #2966: Cannot install velero with go get + go install
**What steps did you take and what happened:**
```
$ GO111MODULE=on go get -v -u github.com/vmware-tanzu/velero
go get: github.com/vmware-tanzu/velero@v1.5.1 requires
        github.com/kubernetes-csi/external-snapshotter/v2@v2.2.0-rc1 requires
        k8s.io/kubernetes@v1.18.0 requires
        k8s.io/api@v0.0.0: reading k8s.io/api/go.mod at revision v0.0.0: unknown revision v0.0.0

$ cd $(go env GOPATH)/src/github.com/vmware-tanzu/velero
$ go install .
go: github.com/kubernetes-csi/external-snapshotter/v2@v2.2.0-rc1 requires
        k8s.io/kubernetes@v1.18.0 requires
        k8s.io/kubectl@v0.0.0 requires
        vbom.ml/util@v0.0.0-20160121211510-db5cfe13f5cc: unrecognized import path "vbom.ml/util": https fetch: Get "https://vbom.ml/util?go-get=1": dial tcp: lookup vbom.ml on XXXXXXXX:53: no such host
```

**What did you expect to happen:**
that the velero cli will be installed successfully

**Environment:**
running inside a fedora container with go installed

- OS (e.g. from `/etc/os-release`):
> $ cat /etc/os-release
> NAME=Fedora
> VERSION="32 (Container Image)"
> ID=fedora
> VERSION_ID=32
> VERSION_CODENAME=""
> PLATFORM_ID="platform:f32"
> PRETTY_NAME="Fedora 32 (Container Image)"
> ANSI_COLOR="0;34"
> LOGO=fedora-logo-icon
> CPE_NAME="cpe:/o:fedoraproject:fedora:32"
> HOME_URL="https://fedoraproject.org/"
> DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f32/system-administrators-guide/"
> SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
> BUG_REPORT_URL="https://bugzilla.redhat.com/"
> REDHAT_BUGZILLA_PRODUCT="Fedora"
> REDHAT_BUGZILLA_PRODUCT_VERSION=32
> REDHAT_SUPPORT_PRODUCT="Fedora"
> REDHAT_SUPPORT_PRODUCT_VERSION=32
> PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
> VARIANT="Container Image"
> VARIANT_ID=container

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

### PR DESCRIPTION
**Pull request by georgettica**
fixes #2966

### CODE REVIEW

**Comment by georgettica** [comment]
yay! all checks passed :) now I need to find a way to test this did what is was supposed to..

suggestions?

File: go.mod
Comment (georgettica) [inline go.mod:38]: not sure why some removed the `-rc2` comment, I will check it later

File: go.mod
Comment (georgettica) [inline go.mod:49]: didn't change replaces yet, but I can if it will help

**Review by zubron — CHANGES_REQUESTED** [review]
Thanks for looking at this! I think the `replace` directives will need to be updated. It looks like those directives were introduced to work around the fact that the CSI snapshotter library imported the Kubernetes libraries directly (see #2651). It may be the case that we can remove them with this change.

Either way, I think there will be more changes needed to this PR as v0.19.0 version isn't being used yet and once it is, it will require other libraries to be updated and changes in the Velero codebase.

File: go.mod
Comment (georgettica) [reply go.mod:49]: Thanks!
I'll chase them down and see if there is an issue regarding this

**Comment by georgettica** [comment]
After having a slack talk I have some bad news..

One of the dependant libraries we need I. Order to make this change are planning to upgrade in Feb


If someone has better insight on it I would gladly push our upgrade to a nearer target
https://kubernetes.slack.com/archives/C8TSNPY4T/p1609798576313300?thread_ts=1609798576.313300&cid=C8TSNPY4T

**Comment by zubron** [comment]
@georgettica Thanks so much for reaching out to the Cluster API folks and checking the status of their releases! :smile: 

Summarising the thread here: It looks like the release with this change will be their next minor release (v0.4.0) and that likely won't be available until end of Q1. Given that we just import it to use the [patch helper](https://github.com/vmware-tanzu/velero/blob/12d4aff3db4f513220ced4b5eca886fb1c2b0823/pkg/controller/backup_storage_location_controller.go#L29) they are considering making it available in controller-runtime but that will require refactoring to make it less specific to cluster API.

@vmware-tanzu/velero-maintainers: It seems our main options are to use an unreleased version of the library or hold off until the official release is available. We already use pre-release versions of other libraries so I think it should be fine to do the same here?

**Comment by jenting** [comment]
I'm 👍  to use the pre-release version.

**Comment by georgettica** [comment]
I'll look into changing the PR in accordance with your changes.

I am glad to reach to the community when I can XD

**Comment by georgettica** [comment]
so with the help of ya'll I moved it forward a bit more..

there is a build issue, caused by this change in controller-runtime
https://github.com/kubernetes-sigs/controller-runtime/commit/8d455073d26bf0b020e3b8cbb9eb9545a336b7f1

I am not sure what to change the function to, so I'll let the sages here look at it

File: go.mod
Comment (georgettica) [reply go.mod]: anyways, they auto-bumped to a newer version so that's ok I guess

File: go.mod
Comment (georgettica) [reply go.mod]: removed all of the replaces and made it build worthy

**Comment by jenting** [comment]
> so with the help of ya'll I moved it forward a bit more..
> 
> there is a build issue, caused by this change in controller-runtime
> [kubernetes-sigs/controller-runtime@8d45507](https://github.com/kubernetes-sigs/controller-runtime/commit/8d455073d26bf0b020e3b8cbb9eb9545a336b7f1)
> 
> I am not sure what to change the function to, so I'll let the sages here look at it

@carlisia might know how to change the function to fulfill controller-runtime API change.
The build fails at https://github.com/vmware-tanzu/velero/blob/4048c02/internal/util/managercontroller/managercontroller.go#L32-L53

**Comment by nrb** [comment]
Based on https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/manager/manager.go#L259, it looks like the `Runnable` interface now takes a `Context` rather than a `<-chan struct{}` channel.

I made the following change locally to the `server.go` and `managercontroller.go` files and was able to compile.

I tried to push this commit to George's fork, but I don't have permissions. I've included it here as a patch file, and it can be applied with `git am < <filename>`, if you want to use it @georgettica.

Thank you tons for digging in to this, I appreciate it!

```patch
From e0056830f1eaef8362d5e382b10afe372bf070ba Mon Sep 17 00:00:00 2001
From: Nolan Brubaker <brubakern@vmware.com>
Date: Thu, 21 Jan 2021 21:19:15 -0500
Subject: [PATCH] Use Context directly on the Runnable interface

New versions of controller-runtime use a Context directly rather than a
channel.

Signed-off-by: Nolan Brubaker <brubakern@vmware.com>
---
 .../managercontroller/managercontroller.go    | 21 ++-----------------
 pkg/cmd/server/server.go                      |  2 +-
 2 files changed, 3 insertions(+), 20 deletions(-)

diff --git a/internal/util/managercontroller/managercontroller.go b/internal/util/managercontroller/managercontroller.go
index 5a009d7f..be7391e2 100644
--- a/internal/util/managercontroller/managercontroller.go
+++ b/internal/util/managercontroller/managercontroller.go
@@ -29,25 +29,8 @@ import (
 // Runnable will turn a "regular" runnable component (such as a controller)
 // into a controller-runtime Runnable
 func Runnable(p controller.Interface, numWorkers int) manager.Runnable {
-	f := func(stop <-chan struct{}) error {
-
-		// Create a cancel context for handling the stop signal.
-		ctx, cancel := context.WithCancel(context.Background())
-		defer cancel()
-
-		// If a signal is received on the stop channel, cancel the
-		// context. This will propagate the cancel into the p.Run
-		// function below.
-		go func() {
-			select {
-			case <-stop:
-				cancel()
-			case <-ctx.Done():
-			}
-		}()
-
-		// This is a blocking call that either completes
-		// or is cancellable on receiving a stop signal.
+	// Pass the provided Context down to the run function.
+	f := func(ctx context.Context) error {
 		return p.Run(ctx, numWorkers)
 	}
 	return manager.RunnableFunc(f)
diff --git a/pkg/cmd/server/server.go b/pkg/cmd/server/server.go
index a84c9a67..5cee3900 100644
--- a/pkg/cmd/server/server.go
+++ b/pkg/cmd/server/server.go
@@ -871,7 +871,7 @@ func (s *server) runControllers(defaultVolumeSnapshotLocations map[string]string
 
 	s.logger.Info("Server starting...")
 
-	if err := s.mgr.Start(s.ctx.Done()); err != nil {
+	if err := s.mgr.Start(s.ctx); err != nil {
 		s.logger.Fatal("Problem starting manager", err)
 	}
 
-- 
2.24.0
```

**Comment by nrb** [comment]
@georgettica Also, I see the DCO check is failing - https://github.com/vmware-tanzu/velero/pull/3202/checks?check_run_id=1661384030. Only the 3 latest commits need the signoff from what I can tell.

**Comment by georgettica** [comment]
Saw it now, and sad you didn't have permissions to push :(

Will look at it on Sunday and will for sure put these changes in.

Thanks!!

**Comment by arianitu** [comment]
Hey @georgettica , just checking on on this PR. It's blocking our builds since we use the velero code directly, any chance we can get this merged?

**Comment by georgettica** [comment]
@arianitu Not yet, as the build is broken.
But there is a fix to solve the next issue I had in line.

I am not working on this full time and will update when I get to a new breakthrough.

**Comment by georgettica** [comment]
all checks are now passing!
now I need to fix all of this mess I did with my commits and we can go forward

**Comment by georgettica** [comment]
tests are failing.. will see what is the matter there next

**Comment by georgettica** [comment]
after some git-fu, all the commits are now on `uptream/main` and I will fix the tests

**Comment by georgettica** [comment]
@zubron I think it works now.. right?

**Comment by georgettica** [comment]
the lint check [here](https://github.com/vmware-tanzu/velero/pull/3202/checks?check_run_id=1768690274) is the issue I have now

File: config/crd/bases/velero.io_restores.yaml
Comment (jenting) [reply config/crd/bases/velero.io_restores.yaml:310]: @georgettica I think your PR is not related to this so this file should not be modified.

Comment (zubron) [reply config/crd/bases/velero.io_restores.yaml:310]: These changes are fine. The `Restore` type includes `v1.Container` and the changes in this file represent changes in the v0.19.2 release.

File: go.mod
Comment (zubron) [inline go.mod:37]: While we're upgrading, do you think we could go for the most recent version of these libraries which is v0.19.7?

File: pkg/controller/backup_storage_location_controller.go
Comment (zubron) [inline pkg/controller/backup_storage_location_controller.go:55]: We can address this in a follow up PR, but I think that now that `Reconcile` takes a context, we can upgrade our existing controllers to not have their [own context](https://github.com/vmware-tanzu/velero/blob/main/pkg/controller/backup_storage_location_controller.go#L41) but instead use this passed in context. Looking at the code from controller-runtime, this is the context that is now passed in to `s.mgr.Start` in `server.go`. I'll open an issue to address this.

**Review by zubron — CHANGES_REQUESTED** [review]
Thanks so much for your continued efforts on this, @georgettica!

This is looking great so far. There is only one further change needed which is to fix the controller tests. As the controller runtime manager now takes a context, we should update this test environment fixture to [use a context](https://github.com/vmware-tanzu/velero/blob/main/pkg/controller/suite_test.go#L87) and then pass that context into [`t.Manager.Start`](https://github.com/vmware-tanzu/velero/blob/main/pkg/controller/suite_test.go#L122). I think it should be enough then to just mark that context as `Done` in the [`stop`](https://github.com/vmware-tanzu/velero/blob/main/pkg/controller/suite_test.go#L126) function.

I also suggested upgrading the Kubernetes libraries to the latest available patch version but if that turns out to be a lot of effort, you can leave them at the version you're currently using and we'll address the upgrade in #3340.

File: pkg/controller/backup_storage_location_controller.go
Comment (zubron) [reply pkg/controller/backup_storage_location_controller.go:55]: Created https://github.com/vmware-tanzu/velero/issues/3353

File: go.mod
Comment (georgettica) [reply go.mod:37]: I can, I just rebased my old changes, will bump now :)

**Comment by georgettica** [comment]
I think I have a solution, hope it works as I expect

**Comment by georgettica** [comment]
EEEverything passes ❤️ 

now it's just a question of what other changes should be tested

**Review by zubron — APPROVED** [review]
Thanks for making the fixes!  LGTM :tada: 

For those interested, I tested this out by creating a small Go project which imports the Velero package. When including `require github.com/vmware-tanzu/velero v1.5.3`, I get the error about the missing version: `k8s.io/api@v0.0.0: reading k8s.io/api/go.mod at revision v0.0.0: unknown revision v0.0.0`. If I add a replace directive to use the branch from this PR (`replace github.com/vmware-tanzu/velero => github.com/georgettica/velero v1.5.2-0.20210126184033-cba96280fdc3`), it builds correctly without error.

**Comment by georgettica** [comment]
Yes!! Thanks :)

[Non-blocker] Can this go.mod be incorporated I to the repo u til everyone has transitioned to working with velero directly?

**Comment by zubron** [comment]
> [Non-blocker] Can this go.mod be incorporated I to the repo u til everyone has transitioned to working with velero directly?

Sorry, I'm not sure what you mean by this? Once this PR is merged, it will be available to use, although it won't be an official release. We're working on v1.6 at the moment which will include this fix, so it will be available as an official release then. If people want to use it before that, then they can use versions from the main branch.

**Review by jenting — APPROVED** [review]
LGTM, thank you.

**Pull request merged**
